Convert `i` commands to rzshell

ret2libc opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [~] I've added tests that prove my fix is effective or that my feature works (if possible)
- [-] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)
- [x] Run on dist-cmd-info-api branch for additional CI: see https://github.com/rizinorg/rizin/pull/1581 , everything is alright except netbsd.. not sure if it's related though.

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

Convert `i` commands to rzshell. In the process, move most of the output to RzTable for consistency.


**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

CI green + a lot of manual testing to ensure stuff still works and are displayed well.
Make sure the issues listed below were really addressed:
- [x] https://github.com/rizinorg/rizin/issues/1516
- [ ] ~~https://github.com/rizinorg/rizin/issues/1514~~ Not done here. `izzz` code is quite messy and would need a separate PR.
- [ ] ~~https://github.com/rizinorg/rizin/issues/1475~~ Not done here. It will be done in a separate PR due to this involving opening a new rzbinfile and operating on that. Might be tricky because API is still not awesome.
- [x] https://github.com/rizinorg/rizin/issues/786
- [x] https://github.com/rizinorg/rizin/issues/1120
- [x] Make sure RZ_BIN_PREFIX is correctly working
- [x] Old code has been removed
- [x] Show N/A when things are not available

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

Closes https://github.com/rizinorg/rizin/issues/1516
Closes https://github.com/rizinorg/rizin/issues/786
Closes https://github.com/rizinorg/rizin/issues/1120
Closes https://github.com/rizinorg/rizin/issues/1027

ret2libc committed:
Start to create some RzBinObject specific API

Adding some unit tests

Fix for new relocs storage system

Fix test_bin unit test for oob-write

ret2libc committed:
Add iA command to newshell

Add ie/iee commands

Add support for is/iE commands

Add support for ii command

Add support for il command

Add support for iM command

Add support for ir command

Add support for iS/iSS commands

Add support for iz command

Add support for i command

Add support for ic command

Add support for iC command

Add support for ih and iHH commands

Add support for iI command

Add support for iL command

Fix descriptions of commands already converted

Add support for idp commands

Add support for iD command

Add support for im commands

Add support for iR command

Add support for it command

Add support for iV commands

Add support for iw command

Add support for ix commands

Add support for iZ command

Add support for ik command

Add support for is. command

Add support for iS. command

Reduce arguments in print_arch

Few doxygen and ownership annotations

Add support for iS= command

Add support for izz and izzz commands

Fix clang-format errors

Few other clang-format fixes

Fix isAnExport comment and check for NULL in rz_core_bin_relocs_print

Add support for ib command

Add support for ic* command

Put section flags as last field of the table

ret2libc commented on librz/core/core_private.h:
```diff
@@ -96,6 +96,12 @@ RZ_IPI void rz_core_agraph_print_json(RzCore *core);
 RZ_IPI void rz_core_agraph_print_write(RzCore *core, const char *filename);
 
+/* cbin.c */
+RZ_IPI void rz_core_bin_entries_print(RzCore *core, RzCmdStateOutput *state);
```
@thestr4ng3r @XVilka @wargio  not sure about these API... shall they get the RzBinFile as argument or shall they just work on the current binfile?

XVilka replied:
But there are not APIs, they are internal. Personally, I think it's fine. Lets see what others think, especially @thestr4ng3r

ret2libc replied:
Probably many of these can just be removed and directly implemented in the cmd_handler in cmd_info.c, what do you think? They don't do much in terms of API..

XVilka replied:
I am completely fine with this proposition. 👍 It makes an absolute sense, especially if not used anywhere else.

ret2libc replied:
Improved the situation here. Marking as resolved.

XVilka commented on librz/bin/bobj.c:
```diff
@@ -685,10 +685,148 @@ RZ_API ut64 rz_bin_object_get_vaddr(RzBinObject *o, ut64 paddr, ut64 vaddr) {
+	return NULL;
+}
+
+RZ_API const RzList *rz_bin_object_reset_strings(RzBin *bin, RzBinFile *bf, RzBinObject *obj) {
```
Please add Doxygen docs for at least non-trivial API functions.

XVilka commented:
Could you also please check if there could be API used for those calls from https://github.com/rizinorg/rizin/issues/1223?
```
librz/main/rz-find.c
234:            rz_sys_cmdf("rz-bin -q%szzz \"%s\"", ro->json ? "j" : "", efile);

librz/egg/egg_Cfile.c
289:    output = rz_sys_cmd_strf("rz-bin -o '%s.text' -O d/S/'%s' '%s'.o",

librz/core/cmd_info.c
703:                            rz_sys_cmdf("rz-bin -O \"%s\" \"%s\"", rz_str_trim_head_ro(input + 1), desc->name);
706:                            rz_sys_cmdf("rz-bin -O help");

librz/core/cmd_debug.c
1673:                           res = rz_sys_cmd_strf("env RZ_BIN_PREFIX=\"%s\" rz-bin %s-B 0x%08" PFMT64x " -S \"%s\" | grep \"%s\"", name, mode, baddr, filesc, sect);
1676:                           res = rz_sys_cmd_strf("env RZ_BIN_PREFIX=\"%s\" rz-bin %s-B 0x%08" PFMT64x " -S \"%s\"", name, mode, baddr, filesc);

```

XVilka commented on librz/core/cbin.c:
```diff
@@ -4771,3 +5354,85 @@ RZ_API RzCmdStatus rz_core_bin_plugins_print(RzBin *bin, RzCmdStateOutput *state
+	}
+}
+
+RZ_API void rz_core_bin_archs_print(RzBin *bin, RzCmdStateOutput *state) {
```
Why some of the printing functions are `RZ_IPI` and this one `RZ_API`? Ideally, I think, we should not have printing functions in the API at all - it should be either passed as `char *` or more complex `RzBuffer`, etc to the API consumer, be it Rizin's own internals or some external library/bindings.

ret2libc replied:
That's because this is used also by rz-bin when doing rz-bin -A

XVilka commented on librz/core/cmd.c:
```diff
@@ -6553,6 +6553,7 @@ RZ_API char *rz_core_cmd_str(RzCore *core, const char *cmd) {
 	rz_cons_push();
 	if (rz_core_cmd(core, cmd, 0) == -1) {
 		//eprintf ("Invalid command: %s\n", cmd);
+		rz_cons_pop();
```
What about extracting this change into the separate PR and merge it right away?

ret2libc replied:
yeah that was the plan i think :D

XVilka commented on librz/util/table.c:
```diff
@@ -195,53 +195,84 @@ RZ_API void rz_table_set_columnsf(RzTable *t, const char *fmt, ...) {
+/**
+ * Add some columns values to the last created row, if any, or create a new row otherwise.
+ */
+RZ_API void rz_table_add_row_columnsf(RzTable *t, const char *fmt, ...) {
```
What if `fmt` is NULL? Please add corresponding check

wargio replied:
fmt must be checked. we will have warnings on this when compiles

XVilka replied:
Do you mean `RZ_PRINTF_CHECK()` macro? It will not work since the table `fmt` is different from `printf()`-like formats.

wargio replied:
no. if you pass to any printf related function a possible null ptr format, then the compiler will complain.

XVilka commented on librz/util/table.c:
```diff
@@ -195,53 +195,84 @@ RZ_API void rz_table_set_columnsf(RzTable *t, const char *fmt, ...) {
+ * Add a new row with the specified columns values.
+ */
 RZ_API void rz_table_add_rowf(RzTable *t, const char *fmt, ...) {
 	va_list ap;
```
Same here

XVilka commented on librz/core/cmd_api.c:
```diff
@@ -330,6 +340,25 @@ RZ_API RzCmdDesc *rz_cmd_get_desc(RzCmd *cmd, const char *cmd_identifier) {
+ * Command descriptors that support multiple output modes can also have a
+ * default one. This function can be used to set it.
+ */
+RZ_API void rz_cmd_desc_set_default_mode(RzCmdDesc *cd, RzOutputMode mode) {
```
Add `RZ_NONNULL RzCmdDesc`

ret2libc replied:
I'm going to skip this because all rz_cmd.h api do not have any annotations. it is worth doing it in a separate PR in one shot.

XVilka commented on librz/core/cmd_api.c:
```diff
@@ -330,6 +340,25 @@ RZ_API RzCmdDesc *rz_cmd_get_desc(RzCmd *cmd, const char *cmd_identifier) {
+	case RZ_CMD_DESC_TYPE_ARGV_STATE:
+		cd->d.argv_state_data.default_mode = mode;
+		break;
+	default:
```
Maybe `rz_warn_if_reached()`?

ret2libc replied:
No, I think it is fine without it, so the caller of this function does not have to know the type of the cd but it can always call the function without having errors.

XVilka commented on librz/core/cmd_api.c:
```diff
@@ -79,6 +79,16 @@ RZ_IPI int rz_output_mode_to_char(RzOutputMode mode) {
 	return -1;
 }
 
+RZ_IPI const char *rz_output_mode_to_summary(RzOutputMode mode) {
```
`RZ_BORROW` here?

XVilka commented on librz/core/cbin.c:
```diff
@@ -4771,3 +5505,85 @@ RZ_API RzCmdStatus rz_core_bin_plugins_print(RzBin *bin, RzCmdStateOutput *state
 	return RZ_CMD_STATUS_OK;
 }
+
+static void print_arch(RzBin *bin, RzCmdStateOutput *state, int num, ut64 offset, ut64 size,
```
Way too many arguments, what about grouping some in some kind of context? `arch`, `bits`, `machine` are good candidates, for example.

XVilka commented on librz/core/cbin.c:
```diff
@@ -2305,6 +2305,748 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+}
+
+RZ_IPI void rz_core_bin_strings_print(RzCore *core, RzCmdStateOutput *state) {
+	bool b64str = rz_config_get_i(core->config, "bin.b64str");
```
What if `state` is NULL? Please add corresponding checks

XVilka commented on librz/core/cbin.c:
```diff
@@ -2305,6 +2305,748 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+}
+
+RZ_IPI void rz_core_bin_segments_print(RzCore *core, RzCmdStateOutput *state, RzList *hashes) {
+	RzBinFile *bf = rz_bin_cur(core->bin);
```
Maybe a quick exit if `hashes` is empty?

ret2libc replied:
no if there are no `hashes` we just skip them, but you still print sections/segments.

XVilka commented on librz/core/cbin.c:
```diff
@@ -2305,6 +2305,748 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+}
+
+RZ_IPI void rz_core_bin_sections_print(RzCore *core, RzCmdStateOutput *state, RzList *hashes) {
+	RzBinFile *bf = rz_bin_cur(core->bin);
```
Maybe a quick exit if `hashes` is empty?

ret2libc replied:
no if there are no hashes we just skip them, but you still print sections/segments.

XVilka commented on librz/core/cbin.c:
```diff
@@ -2305,6 +2305,748 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+}
+
+static void sections_print_table(RzCore *core, RzTable *t, RzBinObject *o, RzBinSection *section, RzList *hashes) {
+	ut64 addr = get_section_addr(core, o, section);
```
Maybe a quick exit if `hashes` is empty?

ret2libc replied:
no if there are no hashes we just skip them, but you still print sections/segments.

XVilka commented on librz/core/cbin.c:
```diff
@@ -2305,6 +2305,748 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+}
+
+static void sections_print_json(RzCore *core, PJ *pj, RzBinObject *o, RzBinSection *section, RzList *hashes) {
+	ut64 addr = get_section_addr(core, o, section);
```
same here

XVilka commented on librz/core/cbin.c:
```diff
@@ -2305,6 +2305,748 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+}
+
+RZ_IPI void rz_core_bin_relocs_print(RzCore *core, RzCmdStateOutput *state) {
+	RzBinFile *bf = rz_bin_cur(core->bin);
```
Please check for `state` not being NULL here.

XVilka commented on librz/core/cbin.c:
```diff
@@ -2305,6 +2305,748 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+}
+
+static bool isAnExport(RzBinSymbol *s) {
+	/* workaround for some bin plugs */
```
`s/plugs/plugins/`, maybe also `s/bin/RzBin/`

XVilka replied:
Why resolved? It wasn't

ret2libc replied:
Just because this is old code i haven't even touched afaik, it was just moved.

XVilka reviewed (changes requested):
A few code style nitpicks

XVilka commented:
Could you please address this warning from `pylint`:
```
************* Module cmd_descs_generate
librz/core/cmd_descs/cmd_descs_generate.py:798:0: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
librz/core/cmd_descs/cmd_descs_generate.py:800:4: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
librz/core/cmd_descs/cmd_descs_generate.py:810:0: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
librz/core/cmd_descs/cmd_descs_generate.py:812:4: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
************* Module check_meson_subproject
sys/check_meson_subproject.py:47:20: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
```

XVilka commented on librz/bin/bobj.c:
```diff
@@ -685,10 +685,148 @@ RZ_API ut64 rz_bin_object_get_vaddr(RzBinObject *o, ut64 paddr, ut64 vaddr) {
+	return obj->strings;
+}
+
+RZ_API bool rz_bin_object_is_string(RzBinObject *obj, ut64 va) {
```
The name a bit weird since it doesn't mean the *whole* object is a string. Just it has a string at the start of it. cc @thestr4ng3r

XVilka replied:
@thestr4ng3r  what do you think?

XVilka replied:
@thestr4ng3r ping?

XVilka commented on librz/bin/bfile.c:
```diff
@@ -860,10 +860,9 @@ static inline bool add_file_hash(RzMsgDigest *md, const char *name, RzList *list
 
-RZ_API RzList *rz_bin_file_compute_hashes(RzBin *bin, ut64 limit) {
-	rz_return_val_if_fail(bin && bin->cur && bin->cur->o, NULL);
+RZ_API RzList *rz_bin_file_compute_hashes(RzBin *bin, RzBinFile *bf, ut64 limit) {
```
Please add Doxygen here. Also the ownership of the return value.

ret2libc replied:
I'd leave that for a separate refactoring because these functions were already there and the PR is already quite huge with all the commands conversions to newshell. I would avoid adding other changes not strictly related to that.

XVilka replied:
Well, the change is minimal, and it's comments. I don't think it should be in a separate PR. You changed the function signature anyway.

XVilka commented on librz/bin/bin.c:
```diff
@@ -594,37 +594,37 @@ RZ_API void rz_bin_set_baddr(RzBin *bin, ut64 baddr) {
 RZ_API RzList *rz_bin_get_entries(RzBin *bin) {
```
Same here - ownership of the return value for all these functions.

XVilka commented on librz/bin/bin.c:
```diff
@@ -648,81 +648,48 @@ RZ_API RzBinSection *rz_bin_get_section_at(RzBinObject *o, ut64 off, int va) {
 }
 
 RZ_API RzList *rz_bin_reset_strings(RzBin *bin) {
```
Ownership mark.

XVilka commented on librz/bin/bin.c:
```diff
@@ -648,81 +648,48 @@ RZ_API RzBinSection *rz_bin_get_section_at(RzBinObject *o, ut64 off, int va) {
+	return (RzList *)rz_bin_object_reset_strings(bin, bf, bf->o);
 }
 
 RZ_API RzList *rz_bin_get_strings(RzBin *bin) {
```
Ownership mark.

XVilka commented on librz/bin/bin.c:
```diff
@@ -648,81 +648,48 @@ RZ_API RzBinSection *rz_bin_get_section_at(RzBinObject *o, ut64 off, int va) {
+	return o ? rz_bin_object_is_string(o, va) : false;
 }
 
 RZ_API RzList *rz_bin_get_symbols(RzBin *bin) {
```
Ownership mark, etc

XVilka commented on librz/bin/bin.c:
```diff
@@ -648,81 +648,48 @@ RZ_API RzBinSection *rz_bin_get_section_at(RzBinObject *o, ut64 off, int va) {
+	return o ? rz_bin_object_is_big_endian(o) : false;
 }
 
 RZ_API int rz_bin_is_static(RzBin *bin) {
```
Convert to a `bool` then.

ret2libc replied:
same, let's do these changes in a separate refactoring just to bin api. this PR is mainly for the conversion of `i` commands to newshell.

XVilka commented on librz/bin/bin.c:
```diff
@@ -1466,3 +1239,45 @@ RZ_API RzBinTrycatch *rz_bin_trycatch_new(ut64 source, ut64 from, ut64 to, ut64
 	free(tc);
 }
+
+RZ_API const RzBinPlugin *rz_bin_plugin_get(RzBin *bin, const char *name) {
```
Add Doxygen, `RZ_NONNULL` for `name`, ownership mark for the return value.

ret2libc replied:
I think the `const` specifier is already enough for the return value. What do you think? When you have `const <type>` it is already clear that the caller does not have to free the object (he can't, because the _free function needs a non-const pointer).

XVilka replied:
Ok, fine. But please add `RZ_NONNULL` for `name` then.

ret2libc replied:
Header file.

XVilka commented on librz/bin/bin.c:
```diff
@@ -1466,3 +1239,45 @@ RZ_API RzBinTrycatch *rz_bin_trycatch_new(ut64 source, ut64 from, ut64 to, ut64
+	return NULL;
+}
+
+RZ_API const RzBinXtrPlugin *rz_bin_xtrplugin_get(RzBin *bin, const char *name) {
```
Add Doxygen, `RZ_NONNULL` for `name`, ownership mark for the return value.

XVilka replied:
Still no `RZ_NONNULL`

ret2libc replied:
It's in the header file.

XVilka commented on librz/bin/bin.c:
```diff
@@ -1466,3 +1239,45 @@ RZ_API RzBinTrycatch *rz_bin_trycatch_new(ut64 source, ut64 from, ut64 to, ut64
+	return NULL;
+}
+
+RZ_API const RzBinLdrPlugin *rz_bin_ldrplugin_get(RzBin *bin, const char *name) {
```
Add Doxygen, `RZ_NONNULL` for `name`, ownership mark for the return value.

XVilka replied:
Still no `RZ_NONNULL`

ret2libc replied:
It's in the header file.

XVilka commented on librz/bin/bobj.c:
```diff
@@ -685,10 +685,189 @@ RZ_API ut64 rz_bin_object_get_vaddr(RzBinObject *o, ut64 paddr, ut64 vaddr) {
 }
 
-RZ_API RzBinAddr *rz_bin_object_get_special_symbol(RzBinObject *o, RzBinSpecialSymbol sym) {
+RZ_API const RzBinAddr *rz_bin_object_get_special_symbol(RzBinObject *o, RzBinSpecialSymbol sym) {
```
Add Doxygen, ownership mark for the return value.

XVilka commented on librz/bin/bobj.c:
```diff
@@ -685,10 +685,189 @@ RZ_API ut64 rz_bin_object_get_vaddr(RzBinObject *o, ut64 paddr, ut64 vaddr) {
+/**
+ * \brief Get list of \p RzBinAddr representing the entry points of the binary object.
+ */
+RZ_API const RzList *rz_bin_object_get_entries(RzBinObject *obj) {
```
Ownership mark, since all of them are borrow

ret2libc replied:
As said above, I think the `const RzList *` already makes it clear that the list is not owned by the caller, no?

XVilka replied:
Well, technically you are right, but then we have 2 different ways of doing the same. It isn't good probably. I wonder if we should just define `RZ_BORROW` as `const`.

ret2libc replied:
I think RZ_BORROW is especially useful when you don't have something `const` and you can't make it const for whatever reason, but you still want the caller to not own the object.

XVilka commented on librz/bin/bobj.c:
```diff
@@ -685,10 +685,189 @@ RZ_API ut64 rz_bin_object_get_vaddr(RzBinObject *o, ut64 paddr, ut64 vaddr) {
+/**
+ * \brief Get list of \p RzBinSection representing only the sections of the binary object.
+ */
+RZ_API RzList *rz_bin_object_get_sections(RzBinObject *obj) {
```
Ownership

wargio commented:
```
[XX] db/abi/platforms/sections MAIN_ABI : arm-linux_gnu_guneabi  sections
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'iS~gold,plt
' bins/abi_bins/elf/platforms/arm-linux-gnueabi-echo
-- stdout
--- expected
+++ actual
@@ -1,2 +1,2 @@
-10  0x00000350   0x50 0x00008350   0x50 -r-- .rel.plt           REL               alloc
-12  0x000003ac   0x8c 0x000083ac   0x8c -r-x .plt               PROGBITS          alloc,execute
+0x00000350 0x50  0x00008350 0x50  -r-- .rel.plt           REL               alloc            0x0
+0x000003ac 0x8c  0x000083ac 0x8c  -r-x .plt               PROGBITS          alloc,execute    0x0
```
Those flags (` alloc,execute`) is actually a list and can become suuuuper long. i highly suggest to always put that in the end to prevent the breakage of the usability and readability of the information.

XVilka commented:
Regarding:
```
[XX] db/cmd/cmd_i iIj
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc iIj~{} bins/elf/analysis/x86-helloworld-gcc
-- stdout
--- expected
+++ actual
@@ -5,11 +5,8 @@
   "bintype": "elf",
   "bits": 32,
   "class": "ELF32",
-  "compiled": "",
   "compiler": "GCC: (GNU) 4.8.2",
-  "dbg_file": "",
   "endian": "LE",
-  "guid": "",
   "intrp": "/lib/ld-linux.so.2",
   "laddr": 0,
   "lang": "c",
@@ -17,7 +14,6 @@
   "maxopsz": 16,
   "minopsz": 1,
   "os": "linux",
-  "cc": "",
   "pcalign": 0,
   "relro": "no",
   "rpath": "NONE",
@@ -33,8 +29,5 @@
   "canary": false,
   "PIE": false,
   "RELROCS": true,
-  "NX": true,
-  "checksums": {
-    
-  }
+  "NX": true
 }
```
I think it's necessary to print empty values still, for the sake of parsers of JSON output
```
[XX] db/cmd/cmd_i ic*
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc ic* bins/elf/libmagic.so
-- stdout
--- expected
+++ actual
@@ -1,111 +1,0 @@
-fs classes
-"f class.JNIEnv = 0x6060"
-"f method._JNIEnv.FindClass_char_const = 0x6060"
-"f method._JNIEnv.DeleteLocalRef__jobject = 0x606a"
-"f method._JNIEnv.NewObject__jclass____jmethodID___... = 0x6074"
-"f method._JNIEnv.GetObjectClass__jobject = 0x608e"
-"f method._JNIEnv.GetMethodID__jclass___char_const___char_const = 0x6098"
-"f method._JNIEnv.CallObjectMethod__jobject____jmethodID___... = 0x60a4"
-"f method._JNIEnv.CallVoidMethod__jobject____jmethodID___... = 0x60c0"
-"f method._JNIEnv.GetStaticMethodID__jclass___char_const___char_const = 0x60dc"
-"f method._JNIEnv.CallStaticObjectMethod__jclass____jmethodID___... = 0x60ea"
-"f method._JNIEnv.CallStaticIntMethod__jclass____jmethodID___... = 0x695e"
-"f method._JNIEnv.CallStaticVoidMethod__jclass____jmethodID___... = 0x697c"
-td "struct _JNIEnv { char empty[0];};"
-"f class.SystemClassLoaderInjector = 0x699c"
```
`ic*` is clearly a bug. In the future we should remove this command, but for now it's necessary still

XVilka commented:
Missing `iAj` output:
```
[XX] db/cmd/cmd_i iA (file mips 64r2)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'iA
iAj
' bins/elf/analysis/mips64r2-busybox-loongson
-- stdout
--- expected
+++ actual
@@ -1,2 +1,4 @@
-0   0x00000000 1039368 mips_64 mips64r2 n64 
-{"bins":[{"arch":"mips","bits":64,"offset":0,"size":1039368,"isa":"mips64r2","features":"n64","machine":"MIPS R3000"}]}
+num offset     size    arch                 machine    
+-------------------------------------------------------
+0   0x00000000 1039368 mips_64 mips64r2 n64 MIPS R3000
```

XVilka commented:
Looks like some of the entries are missing in the `iS` output:
```
[XX] db/abi/platforms/sections MAIN_ABI : x86_64-macho  sections
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'iS~stub
' bins/abi_bins/elf/platforms/x86_64-macho-echo
-- stdout
--- expected
+++ actual
@@ -1,2 +1,2 @@
-1   0x00000f4e   0x1e 0x100000f4e   0x1e -r-x 1.__TEXT.__stubs         SYMBOL_STUBS             PURE_INSTRUCTIONS,SOME_INSTRUCTIONS 0x2
-2   0x00000f6c   0x42 0x100000f6c   0x42 -r-x 2.__TEXT.__stub_helper   REGULAR                  PURE_INSTRUCTIONS,SOME_INSTRUCTIONS 0x4
+0x00000f4e 0x1e  0x100000f4e 0x1e  0x2   -r-x 1.__TEXT.__stubs         SYMBOL_STUBS             PURE_INSTRUCTIONS,SOME_INSTRUCTIONS

[XX] db/abi/platforms/sections MAIN_ABI : x86_64-linux-gnu  sections
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'iS~gold,plt
' bins/abi_bins/elf/platforms/x86_64-linux-gnu-echo
-- stdout
--- expected
+++ actual
@@ -1,3 +1,3 @@
-10  0x00000460   0xd8 0x00400460   0xd8 -r-- .rela.plt          RELA       alloc
-12  0x00000560   0xa0 0x00400560   0xa0 -r-x .plt               PROGBITS   alloc,execute
-23  0x00001000   0x60 0x00601000   0x60 -rw- .got.plt           PROGBITS   write,alloc
+0x00000460 0xd8  0x00400460 0xd8  0x0   -r-- .rela.plt          RELA       alloc
+0x00000560 0xa0  0x00400560 0xa0  0x0   -r-x .plt               PROGBITS   alloc,execute

[XX] db/abi/platforms/sections MAIN_ABI : powerpc32-linux-gnu  sections
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'iS~gold,plt
' bins/abi_bins/elf/platforms/powerpc32-linux-gnu-echo
-- stdout
--- expected
+++ actual
@@ -1,2 +1,2 @@
-10  0x00000350   0x78 0x10000350   0x78 -r-- .rela.plt          RELA     alloc,info
-22  0x00010000   0x28 0x10020000   0x28 -rw- .plt               PROGBITS write,alloc
+0x00000350 0x78  0x10000350 0x78  0x0   -r-- .rela.plt          RELA     alloc,info

[XX] db/abi/platforms/sections MAIN_ABI : powerpc-linux-gnu  sections
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'iS~plt
' bins/abi_bins/elf/platforms/powerpc-linux-gnu-symexec-guess
-- stdout
--- expected
+++ actual
@@ -1,2 +1,2 @@
-10  0x00000294   0x30 0x10000294   0x30 -r-- .rela.plt          RELA     alloc,info
-21  0x00010000   0x10 0x10020000   0x10 -rw- .plt               PROGBITS write,alloc
+0x00000294 0x30  0x10000294 0x30  0x0   -r-- .rela.plt          RELA     alloc,info
```

ret2libc committed:
Start fixing tests

Fix some formats/pe tests

Fix other tests in db/formats and fix assert

wargio committed:
Proper language detection.

wargio committed:
Rewritten blang.c

wargio committed:
implemented icc/icm/icf <class name>

wargio committed:
ignore blocks when printing langs

Fix tests dyldcache, java, pyc

Remove num from iA output

Some other test fixes

XVilka commented on test/db/formats/coff:
```diff
@@ -105,22 +93,17 @@ EXPECT=<<EOF
-
+paddr      size  vaddr      vsize align perm name       type flags 
+-------------------------------------------------------------------
+0x0000017c 0xef  0x00000000 0xef  0x0   ---- .drectve        
```
I am not sure removing the number of section is helpful. Referring to some section by the number is useful for the text description of the Rizin output.

ret2libc replied:
In which case is it useful? I removed those numbers because they weren't coming from the binary, but they were just some indexes incremented by rizin. What do you mean by "text description of the rizin output"?

XVilka replied:
@ret2libc if you put Rizin output in the article or some kind of report. I don't have *strong opinion* about that, just think it is better with numbers.

XVilka commented on test/db/cmd/cmd_i:
```diff
@@ -386,8 +296,10 @@ iA
 EXPECT=<<EOF
-0   0x00000000 266108 mips_32 mips1 o32 
-{"bins":[{"arch":"mips","bits":32,"offset":0,"size":266108,"isa":"mips1","features":"o32","machine":"MIPS R3000"}]}
+num offset     size   arch              machine    
```
For example, you kept the number here.

ret2libc replied:
oh, good catch i should remove this. I was also printing `0` always...

XVilka commented:
Missing one line of output (off-by-one error somewhere?):
```

[XX] db/formats/ar ar file sections
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc iS~:5 ar://bins/ar/libgdbr.a//responses.o
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-1   0x00000040  0x147e 0x08000040  0x147e -r-x .text               PROGBITS alloc,execute

[XX] db/formats/dex DEX dex38.dex ie command
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc ie bins/dex/dex38.dex
-- stdout
--- expected
+++ actual
@@ -1,6 +1,5 @@
-[Entrypoints]
-vaddr=0x00000f70 paddr=0x00000f70 haddr=-1 type=program
-vaddr=0x000010b8 paddr=0x000010b8 haddr=-1 type=program
-vaddr=0x0000121c paddr=0x0000121c haddr=-1 type=program
-
-3 entrypoints
+vaddr      paddr      hvaddr     haddr      type    
+----------------------------------------------------
+0x00000f70 0x00000f70 ---------- ---------- program
+0x000010b8 0x000010b8 ---------- ---------- program

[XX] db/formats/dex DEX HelloWorld.dex imports (ii command)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc ii bins/dex/HelloWorld.dex
-- stdout
--- expected
+++ actual
@@ -1,12 +1,10 @@
-[Imports]
-nth vaddr      bind type  lib       name                                       
--------------------------------------------------------------------------------
+nth vaddr      bind type  lib       name                                               
+---------------------------------------------------------------------------------------
 0   0x00000000 WEAK FIELD java.lang System.out
-1   ---------- WEAK FUNC  java.io   void PrintStream.println(String)
-2   ---------- WEAK FUNC  java.lang void Integer.<init>(int)
-3   ---------- WEAK FUNC  java.lang String Integer.toString()
-4   ---------- WEAK FUNC  java.lang void Object.<init>()
-5   ---------- WEAK FUNC  java.lang void StringBuilder.<init>()
-6   ---------- WEAK FUNC  java.lang StringBuilder StringBuilder.append(String)
-7   ---------- WEAK FUNC  java.lang String StringBuilder.toString()
-
+1   ---------- WEAK FUNC  java.io   PrintStream.sym.imp.void println(String)
+2   ---------- WEAK FUNC  java.lang Integer.sym.imp.void <init>(int)
+3   ---------- WEAK FUNC  java.lang Integer.sym.imp.String toString()
+4   ---------- WEAK FUNC  java.lang Object.sym.imp.void <init>()
+5   ---------- WEAK FUNC  java.lang StringBuilder.sym.imp.void <init>()
+6   ---------- WEAK FUNC  java.lang StringBuilder.sym.imp.StringBuilder append(String)
```

ret2libc commented on librz/bin/blang.c:
```diff
@@ -1,66 +1,50 @@
-// SPDX-FileCopyrightText: 2018-2019 pancake <pancake@nopcode.org>
```
@wargio i think we should keep this as well, WDYT?

XVilka replied:
Yes

wargio replied:
not sure. i completely rewrote it. but yes we can

wargio committed:
Fixed flags on java and added class column

wargio committed:
Fixed alloc issue.

wargio committed:
fixed formats dex/java tests

wargio committed:
Fixed analysis java

wargio committed:
fixed class calls

Fix other tests

Allow optional strings.

Fixing other tests

Add support for ia commands

Fix iCj command

Fix other tests

ret2libc commented:
@wargio please make `classname` arg optional and iterate over all classes on `icf`/`icm` commands when noarg.

XVilka commented on librz/bin/p/bin_dex.c:
```diff
@@ -18,7 +18,7 @@ static RzBinInfo *info(RzBinFile *bf) {
 	}
 
-	binfo->lang = "dalvik";
+	binfo->lang = "java";
```
Can't be DEX file also written in Kotlin or any other JVM language?

XVilka replied:
See this paper for more information about detection: [2105.09591.pdf](https://github.com/rizinorg/rizin/files/7064465/2105.09591.pdf).

wargio replied:
yes, but i will patch with another PR this.

wargio replied:
I have opened an issue to track this: https://github.com/rizinorg/rizin/issues/1537

XVilka commented on librz/core/cbin.c:
```diff
@@ -2047,25 +1754,50 @@ static bool bin_dwarf(RzCore *core, RzBinFile *binfile, PJ *pj, int mode) {
-	rz_return_if_fail(core && s && (mode != RZ_OUTPUT_MODE_JSON || pj));
-	if (mode == RZ_OUTPUT_MODE_JSON) {
+// NOTE: compatibility function, remove as soon as possible
+static bool bin_dwarf_old(RzCore *core, RzBinFile *binfile, PJ *pj, int mode) {
```
Maybe put `RZ_DEPRECATED` before then?

XVilka replied:
Then the compiler can spit the warning when it's properly defined like in https://github.com/rizinorg/rizin/pull/865

XVilka commented on librz/core/cbin.c:
```diff
@@ -2325,990 +2053,1327 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+	if (!bf) {
+		// TODO: manually creating an RzBinFile like this is a hack and abuse of RzBin API
+		// If we don't want to use an RzBinFile for searching strings, the raw strings search
+		// should be refactored out of bin.
```
See https://github.com/rizinorg/rizin/pull/1535

XVilka commented on librz/core/cbin.c:
```diff
@@ -3570,52 +3619,1488 @@ static void classdump_java(RzCore *r, RzBinClass *c) {
+	return ret;
+}
+
+/* Map Sections to Segments https://github.com/rizinorg/rizin/issues/14647 */
```
This link doesn't exist, the issue is from Radare2, moreover it was resolved long time ago in https://github.com/rizinorg/rizin/commit/53f21d9ae6e50cc6a76745829c37c134e50542f4
Just remove the link.

XVilka commented on librz/core/cbin.c:
```diff
@@ -3570,52 +3619,1488 @@ static void classdump_java(RzCore *r, RzBinClass *c) {
+	RzList *sections = rz_list_new();
+	RzList *segments = rz_list_new();
+	RzList *tmp = rz_bin_get_sections(bin);
+	RzTable *table = rz_table_new();
```
Please add NULL checks for lists and table.

XVilka commented on librz/core/cbin.c:
```diff
@@ -3570,52 +3619,1488 @@ static void classdump_java(RzCore *r, RzBinClass *c) {
+	RzBinFile *bf = r->bin->cur;
+	RzBinObject *o = bf ? bf->o : NULL;
+	RzBinPlugin *plugin = rz_bin_file_cur_plugin(bf);
+	rz_return_val_if_fail(table, false);
```
Please convert to the runtime check. Please also rearrange the code so that `bf`/`o` are checked at the beginning as in other places.

XVilka commented on librz/core/cbin.c:
```diff
@@ -4358,13 +5798,13 @@ RZ_API int rz_core_bin_info(RzCore *core, int action, PJ *pj, int mode, int va,
 		}
 	}
 	if ((action & RZ_CORE_BIN_ACC_INFO)) {
```
Just remove this.

XVilka commented on librz/core/cbin.c:
```diff
@@ -4960,3 +6410,229 @@ RZ_API RzCmdStatus rz_core_bin_plugins_print(RzBin *bin, RzCmdStateOutput *state
+	bin_dwarf(core, bf, state);
+}
+
+RZ_IPI char *rz_core_bin_pdb_get_filename(RzCore *core) {
```
Add `RZ_OWN`

XVilka commented on librz/core/cmd_info.c:
```diff
@@ -429,28 +417,74 @@ static bool print_source_info(RzCore *core, PrintSourceInfoType type, RzOutputMo
 	return true;
 }
 
+RZ_IPI int rz_cmd_info_kuery(void *data, const char *input) {
```
What about converting this to the newshell as well, while you are on it?

XVilka commented on librz/core/cmd_info.c:
```diff
@@ -429,28 +417,74 @@ static bool print_source_info(RzCore *core, PrintSourceInfoType type, RzOutputMo
+	rz_core_block_read(core);
+}
+
 RZ_IPI int rz_cmd_info(void *data, const char *input) {
```
Since you are converting everything, just remove this function afterwards.

XVilka commented on librz/util/table.c:
```diff
@@ -102,7 +102,23 @@ RZ_API void rz_table_free(RzTable *t) {
 	free(t);
 }
 
+static bool column_exists(RzTable *t, const char *name) {
```
What about extracting this `RzTable` change into a separate PR, it will help to merge it quicker? cc @ret2libc @wargio

XVilka commented on test/db/cmd/cmd_json:
```diff
@@ -48,6 +48,6 @@ NAME=empty irj should print empty array
 FILE==
 CMDS=irj
 EXPECT=<<EOF
-[]
```
Looks like a regression, no?

ret2libc replied:
I'm not that sure, because the binary info don't exist at all so the commands fail. I think [] means there are no relocations, while in this case there is just no binary at all.

XVilka commented on test/db/cmd/feat_grep:
```diff
@@ -170,15 +171,15 @@ i~:-1
+size    0x400
+file    malloc://1024
+block   0x100
+type    
```
What about printing `N/A` when there is no value available?

XVilka commented on test/db/formats/le:
```diff
@@ -158,7 +155,7 @@ cat $relocs~0x0004e~?
 EOF
 EXPECT=<<EOF
-5241
+5239
```
This looks like regression to me. What exact difference in the output?

XVilka commented on test/db/formats/mdmp:
```diff
@@ -10,25 +10,26 @@ NAME=mdmp info
 bintype  mdmp
 bits     64
 retguard false
+class    
```
Same here. What about `N/A` for `clas` and `lang` here?

XVilka commented on test/db/formats/mdmp:
```diff
@@ -229,28 +225,28 @@ RUN
 CMDS=is~?
 EXPECT=<<EOF
-7469
+7468
```
Why there is a difference?

ret2libc replied:
There was the [symbol] header before.

XVilka commented on test/db/formats/mdmp:
```diff
@@ -274,7 +270,7 @@ NAME=mdmp 32bit - imports count
 CMDS=ii~?
 EXPECT=<<EOF
-2031
+2030
```
Same

XVilka commented on test/db/formats/mdmp:
```diff
@@ -342,16 +338,15 @@ NAME=mdmp 32bit - strings count
 CMDS=iz~?
 EXPECT=<<EOF
-10277
+10276
```
Same

XVilka commented on test/db/formats/mdmp:
```diff
@@ -472,7 +467,7 @@ NAME=mdmp 64bit - strings count
 CMDS=iz~?
 EXPECT=<<EOF
-22901
+22900
```
Also, why there is a difference?

XVilka commented on test/unit/test_table.c:
```diff
@@ -327,10 +327,10 @@ bool test_rz_table_transpose() {
 	rz_table_add_row(t, "d", "100", NULL);
 	char *table = rz_table_tostring(rz_table_transpose(t));
 	mu_assert_streq(table,
-		"Name  Value Value Value Value \n"
```
Same, please extract this change in that separate `RzTable` PR.

XVilka reviewed (changes requested):
Overall, looks really good now.

XVilka commented on librz/core/cbin.c:
```diff
@@ -2325,990 +2053,1327 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+	int bin_demangle = rz_config_get_i(core->config, "bin.demangle");
+	bool keep_lib = rz_config_get_i(core->config, "bin.demangle.libs");
+	int va = (core->io->va || core->bin->is_debugger) ? VA_TRUE : VA_FALSE;
+	RzBinFile *bf = rz_bin_cur(core->bin);
```
Move the check before getting the config variables

XVilka commented on librz/core/cbin.c:
```diff
@@ -2325,990 +2053,1327 @@ static const char *bin_reloc_type_name(RzBinReloc *reloc) {
+		return;
+	}
+
+	const RzList *sections = rz_bin_object_get_sections(o);
```
What if it's NULL? Should there be a quick exit?

ret2libc replied:
Nah, the rz_foreach will just do no iterations, which is fine.

ret2libc committed:
RzBin: add rz_bin_object_sections_mapping_list API

ret2libc committed:
Fix rz-bin for new API

Remove unused va var in rz-bin.c

Fix iA/-A output

XVilka commented on librz/main/rz-bin.c:
```diff
@@ -1070,38 +1164,56 @@ RZ_API int rz_main_rz_bin(int argc, const char **argv) {
+		result = 0;
+		goto err;
 	}
 #define isradjson (rad == RZ_MODE_JSON && actions > 0)
```
Rename this one please as well.

ret2libc replied:
This is existing code i haven't touch, let it be.

ret2libc commented on librz/main/rz-bin.c:
```diff
@@ -9,6 +9,109 @@
+static void sections_mapping_print(RzCore *core, RzCmdStateOutput *state) {
+	rz_cmd_state_output_set_columnsf(state, "ss", "Segment", "Sections");
+	rz_cmd_state_output_array_start(state);
+
```
Implement me

wargio committed:
implemented classdump_cpp

wargio committed:
Support namespaces

ret2libc committed:
Fix rz-bin output

Fix other tests

It should fix everything

Use icc, icl does not exist anymore

Fix archos/darwin test

Fix clang-format on rz-bin.c

Update output of broken tests for icc

Add rz_return_ asserts in new _print functions

ret2libc committed:
Address some PR feedback

ret2libc committed:
Provide rz_core_bin_print API

XVilka commented:
I see you changed the order of information printed. Please update the test as well.
This one looks like a regression though:
```
[XX] db/tools/rz_bin rz-bin -Zj
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '!rz-bin -Zj ${RZ_FILE}' bins/pe/base.exe
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-{"size":216569}
```

ret2libc committed:
Select which parts to print also based on mode

ret2libc committed:
Offset should be UT64_MAX, not 0

ret2libc committed:
Fix tests

ret2libc committed:
Return bool from all rz_core_bin_*_print API

XVilka commented:
There is a fail on Linux with both GCC and TinyCC compilers:
```

[XX] db/archos/linux-x64/dbg_dmi dmi ld
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -d -Qc 'dmiqq ld
dmiqq ld _dl_rtld_di_serinfo
dk 9
' /bin/ls
-- stdout
--- expected
+++ actual
@@ -6,14 +6,15 @@
 _dl_catch_exception
 GLIBC_2.4
 _dl_catch_error
-_dl_starting_up
+free
+realloc
 _dl_allocate_tls
 _r_debug
 __libc_stack_end
-_dl_fatal_printf
 _dl_deallocate_tls
...
```

Moreover, 3 failing tests on NetBSD (but not FreeBSD or OpenBSD):
```
[XX] db/cmd/feat_variables ?p and ?P
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '?p 0xdeadbeef
?P 0xdeadbeef
' bins/elf/analysis/simple.elf
-- stdout
--- expected
+++ actual
@@ -1,2 +1,0 @@
-no map at 0xdeadbeef
-no map at 0xdeadbeef

-- exit status: 
[XX] db/cmd/feat_variables ?p and ?P
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '?p 0x08000060
?P 0x00000060
' bins/elf/analysis/simple.elf
-- stdout
--- expected
+++ actual
@@ -1,2 +1,0 @@
-0x00000060
-0x08000060

-- exit status: 
[XX] db/cmd/feat_yank yank-segfault4
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'y 1000000000
?e ok
' =
-- stdout
--- expected
+++ actual
@@ -1,1 +1,0 @@
-ok

-- exit status: -1
```

XVilka commented on librz/bin/blang.c:
```diff
@@ -1,195 +1,155 @@
 
-static bool check_dlang(RzBinSymbol *sym) {
+static inline bool check_dlang(RzBinSymbol *sym) {
 	if (!strncmp(sym->name, "_D2", 3)) {
```
Couldn't you just make it `if (!strncmp() || !strncmp())` instead?

XVilka commented on librz/bin/blang.c:
```diff
@@ -1,195 +1,155 @@
-	}
-	if (!strncmp(name, "__Z", 3)) {
+static inline bool check_cxx(RzBinSymbol *sym) {
+	if (!strncmp(sym->name, "_Z", 2)) {
```
Same here, you can fold two comparisons together.

XVilka commented on librz/bin/bobj.c:
```diff
@@ -685,10 +683,249 @@ RZ_API ut64 rz_bin_object_get_vaddr(RzBinObject *o, ut64 paddr, ut64 vaddr) {
+
+		rz_list_foreach (sections, iter2, section) {
+			RzInterval section_itv = (RzInterval){ section->vaddr, section->size };
+			if (rz_itv_begin(section_itv) >= rz_itv_begin(segment_itv) && rz_itv_end(section_itv) <= rz_itv_end(segment_itv) && section->name[0]) {
```
Maybe better make it multiline for readability?

XVilka commented on librz/core/cbin.c:
```diff
@@ -447,6 +378,197 @@ RZ_API bool rz_core_bin_apply_all_info(RzCore *r, RzBinFile *binfile) {
+		}
+		rz_cons_printf("%s\n", s);
+		free(s);
+	} else if (state->mode == RZ_OUTPUT_MODE_JSON || state->mode == RZ_OUTPUT_MODE_LONG_JSON) {
```
@ret2libc Is this still required now?

ret2libc replied:
yes it is. it is used to "combine" the json output in `rz_core_bin_print`

XVilka reviewed (changes requested):
@ret2libc @wargio there are two commands that do the same: `iH` and `ih`.
I suggest to move current `ih` output to the `iht` for table and `iH` to the `ih`. What do you think?

ret2libc committed:
Use RzBinFile in new rz_core_bin_*_print API

ret2libc committed:
Restore dbg_dmi test

ret2libc committed:
Remove oldshell handlers for i commands

ret2libc committed:
Remove ieq call in rizin, not needed

file.analyze cfg var does not even exist

XVilka commented:
```
[XX] db/formats/mangling/mangling Very long one with member functions
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '"iD msvc ??_G?$impl@V?$binder2@V?$io_op@V?$basic_stream_socket@Vtcp@ip@asio@boost@@V?$any_executor@U?$context_as_t@AEAVexecution_context@asio@boost@@@execution@asio@boost@@U?$never_t@$0A@@blocking@detail@234@U?$prefer_only@U?$possibly_t@$0A@@blocking@detail@execution@asio@boost@@@234@U?$prefer_only@U?$tracked_t@$0A@@outstanding_work@detail@execution@asio@boost@@@234@U?$prefer_only@U?$untracked_t@$0A@@outstanding_work@detail@execution@asio@boost@@@234@U?$prefer_only@U?$fork_t@$0A@@relationship@detail@execution@asio@boost@@@234@U?$prefer_only@U?$continuation_t@$0A@@relationship@detail@execution@asio@boost@@@234@@execution@34@@asio@boost@@V?$read_op@Vmutable_buffers_1@asio@boost@@@detail@ssl@23@V?$read_op@V?$stream@V?$basic_stream_socket@Vtcp@ip@asio@boost@@V?$any_executor@U?$context_as_t@AEAVexecution_context@asio@boost@@@execution@asio@boost@@U?$never_t@$0A@@blocking@detail@234@U?$prefer_only@U?$possibly_t@$0A@@blocking@detail@execution@asio@boost@@@234@U?$prefer_only@U?$tracked_t@$0A@@outstanding_work@detail@execution@asio@boost@@@234@U?$prefer_only@U?$untracked_t@$0A@@outstanding_work@detail@execution@asio@boost@@@234@U?$prefer_only@U?$fork_t@$0A@@relationship@detail@execution@asio@boost@@@234@U?$prefer_only@U?$continuation_t@$0A@@relationship@detail@execution@asio@boost@@@234@@execution@34@@asio@boost@@@ssl@asio@boost@@Vmutable_buffers_1@34@PEBVmutable_buffer@34@Vtransfer_at_least_t@detail@34@V?$wrapped_handler@Vstrand@io_context@asio@boost@@V?$custom_alloc_handler@V?$_Binder@U_Unforced@std@@P8?$connection@Utransport_config@asio_tls@config@websocketpp@@@asio@transport@websocketpp@@EAAXV?$function@$$A6AXAEBVerror_code@std@@_K@Z@2@AEBVerror_code@system@boost@@_K@ZV?$shared_ptr@V?$connection@Utransport_config@asio_tls@config@websocketpp@@@asio@transport@websocketpp@@@2@AEAV72@AEBU?$_Ph@$00@2@AEBU?$_Ph@$01@2@@std@@@3transport@websocketpp@@Uis_continuation_if_running@detail@34@@834@@523@@detail@ssl@asio@boost@@Verror_code@system@5@_K@detail@asio@boost@@V?$allocator@X@std@@@executor_function@detail@asio@boost@@QEAAPEAXI@Z"' =
-- stdout
--- expected
+++ actual
@@ -1,1 +1,0 @@

[XX] db/cmd/cmd_i i (no rbin - file x86_64)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -n -Qc i~!file bins/elf/analysis/hello-linux-x86_64
-- stdout
--- expected
+++ actual
@@ -1,6 +1,0 @@
-fd      3
-size    0x1a36
-humansz 6.6K
-mode    r-x
-iorw    false

[XX] db/cmd/cmd_i izzj (no rbin - file x86_64)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -n -Qc 'e bin.str.purge=all,!0x2fa,!0x1953
izzj
' bins/elf/analysis/hello-linux-x86_64
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-[{"vaddr":762,"paddr":762,"ordinal":4,"size":5,"length":4,"section":"","type":"ascii","string":"puts"},{"vaddr":6483,"paddr":6483,"ordinal":65,"size":18,"length":17,"section":"","type":"ascii","string":"puts@@GLIBC_2.2.5"}]

[XX] db/cmd/cmd_i izz (no rbin - file x86_64)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -n -Qc izz~puts bins/elf/analysis/hello-linux-x86_64
-- stdout
--- expected
+++ actual
@@ -1,2 +1,0 @@
-4   0x000002fa 0x000002fa 4   5            ascii   puts

[XX] db/cmd/cmd_i i (all) (malloc)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -a x86 -b 64 -Qc 'i
ia
ii
ie
is
iS
iz
' malloc://1025
-- stdout
--- expected
+++ actual
@@ -41,8 +41,8 @@
 -----------------------------
 
 [Entries]
-vaddr paddr hvaddr haddr type 
-------------------------------
+
+
 
 [Exports]
 nth paddr vaddr bind type size lib name 
@@ -70,8 +70,8 @@
 
 nth vaddr bind type lib name 
 -----------------------------
-vaddr paddr hvaddr haddr type 
-------------------------------
+
+
 nth paddr vaddr bind type size lib name 
 ----------------------------------------
 paddr size vaddr vsize align perm name type flags 

WARNING: pj_ks: assertion 'j && k && v' failed (line 170)
```

ret2libc committed:
Allow NULL bf in rz_cmd_info_handler

ret2libc committed:
fix last tests

ret2libc committed:
fix swift rz-bin test

ret2libc committed:
Re-add support for RZ_BIN_PREFIX for symbols/sections/segments/imports/relocs

ret2libc committed:
Move 'ib' command to 'obR'

ret2libc committed:
Show 'N/A' in 'i' command when info are not available

ret2libc committed:
Assert on bf->o, just to be sure

ret2libc commented:
Please ignore the failing tests for now. Just look at the output and see if the changes make sense. It is just a matter of updating tests to use `obR` instead of `ib` and then fix some tests... but please tell me if code and behaviours make sense!

XVilka reviewed (approved):
I like the new output 👍

ret2libc committed:
Fix tests after changes to  output

ret2libc committed:
Make error messages clearer in idp commands

wargio committed:
Merge branch 'dev' into cmd-info-api
